File: libs/ui_widgets/src/facies_eyedropper.cpp
#include "pwb/ui_widgets/facies_eyedropper.hpp"

#include "pwb/ui_widgets/facies_palette_widget.hpp"  // facies_color

namespace pwb::ui_widgets {

QVariantMap facies_pick_result_to_variant(
    const core::FaciesPickResult& result) {
    QVariantMap out;
    out.insert("facies", QString::fromStdString(result.facies));
    out.insert("sub_facies", QString::fromStdString(result.sub_facies));
    out.insert("micro_facies", QString::fromStdString(result.micro_facies));
    out.insert("level", QString::fromStdString(result.level));
    out.insert("layer_id", QString::fromStdString(result.layer_id));
    out.insert("feature_id", QString::fromStdString(result.feature_id));
    out.insert("layer_name", QString::fromStdString(result.layer_name));
    out.insert("color", QString::fromStdString(result.color));
    // Python emits None for unmapped patterns — QVariant() (invalid).
    out.insert("pattern",
               result.pattern
                   ? QVariant(QString::fromStdString(*result.pattern))
                   : QVariant());
    return out;
}

FaciesEyedropper::FaciesEyedropper(QObject* parent) : QObject(parent) {
    color_of_ = [](const std::string& name) {
        return facies_color(QString::fromStdString(name)).toStdString();
    };
}

void FaciesEyedropper::bind(core::FaciesIdentifyFn identify) {
    identify_ = std::move(identify);
}

void FaciesEyedropper::bind_color(core::FaciesColorFn color_of) {
    if (color_of) color_of_ = std::move(color_of);
}

bool FaciesEyedropper::handle_click(double x, double y) {
    if (!active_ || !identify_) return false;
    const auto result = core::pick_facies_at(x, y, identify_, color_of_);
    if (!result) {
        emit pick_missed();
        return false;
    }
    emit picked(facies_pick_result_to_variant(*result));
    return true;
}

}  // namespace pwb::ui_widgets

File: legacy/python_reference/product/paleo_workbench/ui/components/facies_eyedropper.py
"""相带吸色管（M2）——点击地图既有相带，一键吸取属性/颜色/花纹。

数据路径双栈一致（00-decisions D10）：拾取一律走 Python 权威
（``CompositeEditController.identify_all`` 或等价 identify 回调），按
返回序取第一个携带相带属性的命中（identify_all 已按图层栈序返回，
顶层优先）；未命中如实返回 None——不猜、不弹窗。
"""
from __future__ import annotations

from typing import Any, Callable, Iterable

from PySide6.QtCore import QObject, Signal

from paleo_workbench.mapping.facies_taxonomy import FaciesTaxonomy


def pick_facies_at(
    point: tuple[float, float],
    identify: Callable[[tuple[float, float]], Iterable[dict]],
) -> dict | None:
    """地图点 → 相带拾取结果（无相带命中返回 None）。

    返回 dict：facies/sub_facies/micro_facies/level + 来源（layer_id/
    feature_id）+ 分类渲颜色 color + 花纹 pattern（无映射为 None）。
    """
    from paleo_workbench.ui.components.facies_palette_widget import facies_color
    from paleo_workbench.mapping.facies_patterns import pattern_id_for_facies

    for hit in identify(point) or ():
        attributes = dict(hit.get("attributes") or {})
        facies = str(attributes.get("facies") or "").strip()
        if not facies:
            continue  # 非相带要素/图层：跳过取下一命中
        selection = {
            "facies": facies,
            "sub_facies": str(attributes.get("sub_facies") or ""),
            "micro_facies": str(attributes.get("micro_facies") or ""),
        }
        return {
            **selection,
            "level": FaciesTaxonomy.selection_level(selection),
            "layer_id": str(hit.get("layer_id") or ""),
            "feature_id": str(hit.get("feature_id") or ""),
            "layer_name": str(hit.get("layer_name") or ""),
            "color": facies_color(facies),
            "pattern": pattern_id_for_facies(facies),
        }
    return None


class FaciesEyedropper(QObject):
    """吸色管会话状态（激活/点击/装备转发）。

    纯协调器：不拥有画布或工具状态；点击事件由宿主（CompositeDocument）
    在激活期路由到 :meth:`handle_click`。``identify`` 注入返回
    identify_all 形状的结果列表。
    """

    picked = Signal(dict)
    pick_missed = Signal()

    def __init__(self, parent: QObject | None = None) -> None:
        super().__init__(parent)
        self._active = False
        self._identify: Callable[[tuple[float, float]], list[dict]] | None = None

    @property
    def active(self) -> bool:
        return self._active

    def bind(self, identify: Callable[[tuple[float, float]], list[dict]]) -> None:
        self._identify = identify

    def set_active(self, active: bool) -> None:
        self._active = bool(active)

    def handle_click(self, point: tuple[float, float]) -> bool:
        """处理一次地图点击；命中装备（picked），空击保持原装备（missed）。"""
        if not self._active or self._identify is None:
            return False
        result = pick_facies_at(point, self._identify)
        if result is None:
            self.pick_missed.emit()
            return False
        self.picked.emit(result)
        return True
